src/api/dirmonitor/polling.c

```diff
@@ -0,0 +1,123 @@
+#include <stdlib.h>
+#include <string.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <SDL.h>
+
+enum {
+  EVENT_MODIFIED
+};
+
+struct file_info {
+  int fd;
+  struct stat file_stat;
+  int events;
+};
+
+struct dirmonitor_internal {
+  struct file_info* fds;
+  size_t fd_count;
+};
+
+  
+struct dirmonitor_internal* init_dirmonitor() {
+  struct dirmonitor_internal* monitor = calloc(1, sizeof(struct dirmonitor_internal));
+
+  monitor->fd_count = 0;
+  monitor->fds = NULL;
+  return monitor;
+}
+
+void deinit_dirmonitor(struct dirmonitor_internal* monitor) {
+  for (size_t i = 0; i < monitor->fd_count; ++i)
+    close(monitor->fds[i].fd);
+  free(monitor->fds);
+}
+
+int get_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int len)
+{
+  int c = 0;
+  for (size_t i = 0; i < monitor->fd_count; ++i) {
+    if (monitor->fds[i].fd < 0)
+      continue;
+
+    struct stat new_stat;
+    fstat(monitor->fds[i].fd, &new_stat);
+
+    if (new_stat.st_mtime > monitor->fds[i].file_stat.st_mtime)
+    {
+      monitor->fds[i].file_stat = new_stat;
+      monitor->fds[i].events = EVENT_MODIFIED;
+      c += 1;
+    }
+  }
+
+  if (!c)
+    SDL_Delay(100);
+
+  return c;
+}
+
+int translate_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int count, int (*change_callback)(int, const char*, void*), void* data) {
+  for (size_t i = 0; i < monitor->fd_count && count; ++i) {
+    if (!(monitor->fds[i].events & EVENT_MODIFIED)) {
+      --count;
+      monitor->fds[i].events = 0;
+      change_callback(monitor->fds[i].fd, NULL, data);
+    }
+  }
+
+  return count != 0;
+}
+
+int add_dirmonitor(struct dirmonitor_internal* monitor, const char* path)
+{
+  int fd = open(path, O_RDONLY | O_NOFOLLOW);
+  if (fd == -1)
+    return -1;
+
+  // Check if there is any space we can reuse
+  for (size_t i = 0; i < monitor->fd_count; ++i) {
+    if (monitor->fds[i].fd == -1)
+    {
+      monitor->fds[i].fd = fd;
+      fstat(fd, &monitor->fds[i].file_stat);
+      return fd;
+    }
+  }
+
+  // There is no empty space, add a new entry;
+  monitor->fd_count++;
+  struct file_info* old_fds = monitor->fds;
+  struct file_info* new_fds;
+  new_fds = malloc(sizeof(*new_fds) * monitor->fd_count);
```

Comment:
I wonder if it should allocate a bit more to avoid reallocating too much.
But then you'd have to also keep track of the allocated space.

Why not `realloc`? POSIX reasons?

Reply:
sure, I guess?
its already written in a way where it reuses space it no longer needs.
an alternative would be to give it a default size of e.g. 16 entries and then have it expand on-demand

the reason why I'm using malloc instead of realloc is that realloc leaves the new space uninitialized, which will inevitably result in a race condition

Reply:
>the reason why I'm using malloc instead of realloc is that realloc leaves the new space uninitialized

I mean, you're using `memset` anyways.

RE race conditions, should there be a mutex that handles changes to `monitor`, considering that `get_changes_dirmonitor` is running in a different thread?

Reply:
nah, the only places that that modify it are add and delete and on all computers I can imagine this should be possible in one cycle

Reply:
Still kinda iffy, also considering that we're copying entire structs around.

There is also the problem that `monitor->fd_count` changes before `monitor->fds = new_fds;`, so there is potential for oob access.

Reply:
You're right, I'm gonna rework this to get rid of any possible race conditions

Reply:
reworked now:
- allocates a new and increased list without modifying the monitor data
- copies existing data over
- replaces the pointer in the monitor with the new one
- increases the fd count

this should prevent any possible race conditions without having to use a mutex